src/components/modals/PrimaryNameModal/PrimaryNameModal.tsx

```diff
@@ -0,0 +1,316 @@
+import { Loader } from '@src/components/layout';
+import ArweaveID, {
+  ArweaveIdTypes,
+} from '@src/components/layout/ArweaveID/ArweaveID';
+import TransactionCost from '@src/components/layout/TransactionCost/TransactionCost';
+import useDomainInfo from '@src/hooks/useDomainInfo';
+import { usePrimaryName } from '@src/hooks/usePrimaryName';
+import { ArweaveTransactionID } from '@src/services/arweave/ArweaveTransactionID';
+import {
+  useGlobalState,
+  useTransactionState,
+  useWalletState,
+} from '@src/state';
+import dispatchANTInteraction from '@src/state/actions/dispatchANTInteraction';
+import dispatchArIOInteraction from '@src/state/actions/dispatchArIOInteraction';
+import {
+  ANT_INTERACTION_TYPES,
+  ARNS_INTERACTION_TYPES,
+  ContractInteraction,
+  PrimaryNameRequestPayload,
+  RemovePrimaryNamesPayload,
+} from '@src/types';
+import {
+  decodePrimaryName,
+  encodePrimaryName,
+  shortPrimaryName,
+} from '@src/utils';
+import { PRIMARY_NAME_COST } from '@src/utils/constants';
+import eventEmitter from '@src/utils/events';
+import { useQueryClient } from '@tanstack/react-query';
+import { ArrowRight } from 'lucide-react';
+import { useEffect, useState } from 'react';
+
+import DialogModal from '../DialogModal/DialogModal';
+
+/**
+ * Three workflows
+ * - set primary name
+ * use ario contract to create request than approve the request, then verify the primary name
+ * - set primary name when it will change the primary name
+ * use ario contract to create the request then approve the request with the connected ant, then verify the primary name
+ *  should show the comparison between the previous primary name and new primary name
+ * - remove primary name
+ * use the ant to remove the primary name
+ *
+ */
+
+enum PRIMARY_NAME_WORKFLOWS {
+  REQUEST = 'Request Primary Name',
+  CHANGE = 'Change Primary Name',
+  REMOVE = 'Remove Primary Name',
+}
+
+function isPrimaryNameRequest(
+  payload: any,
+): payload is PrimaryNameRequestPayload {
+  return typeof payload?.name == 'string';
+}
+
+function isRemovePrimaryNamesPayload(
+  payload: any,
+): payload is RemovePrimaryNamesPayload {
+  return (
+    typeof payload?.names == 'object' &&
+    Array.isArray(payload.names) &&
+    payload.names.every((s: unknown) => typeof s === 'string') &&
+    typeof payload?.ioProcessId == 'string'
+  );
+}
+
+function PrimaryNameModal({
+  visible,
+  setVisible,
+}: {
+  visible: boolean;
+  setVisible: (visible: boolean) => void;
+}) {
+  const queryClient = useQueryClient();
+  const [{ ioTicker, ioProcessId, arioContract }] = useGlobalState();
+  const [{ wallet, walletAddress }] = useWalletState();
+  const { data: primaryNameData, isLoading: isLoadingPrimaryNameData } =
+    usePrimaryName();
+
+  const [{ transactionData }, dispatchTransactionState] = useTransactionState();
+
+  const transactionPayload = isRemovePrimaryNamesPayload(transactionData)
```

Comment:
Look s a bit awkward, could join the two conditions with an || if they're both going to resolve to transactionData